### ISSUE TITLE
RoBERTa: A Robustly Optimized BERT Pretraining Approach

### ISSUE THREAD

**Issue by fhiyo** [fhiyo/ml_papers#43]
## 0. 論文

Yinhan Liu ∗§, Myle Ott ∗§, Naman Goyal ∗§, Jingfei Du ∗§, Mandar Joshi †
Danqi Chen §, Omer Levy §, Mike Lewis §, Luke Zettlemoyer †§, Veselin Stoyanov §

† Paul G. Allen School of Computer Science & Engineering
§ Facebook AI

## 1. どんなもの？

BERTの実験条件を見直すことでBERTの論文内で示されたのpre-trainingはunderfitであることを主張し、再度pre-trainingを行うことで性能向上を示した論文。

### 新しいデータセットを追加

CommonCrawl Newsデータセットから英語記事の部分を抽出したCC-NEWSデータセットを作成。63M記事、76GB。これをpre-trainingのデータセットとして追加した。

(BERTで用いているBOOKCURPUS + English Wikipediaのサイズは合わせて16GBである)

他にRoBERTaには、OPENWEBTEXT (38GB) とSTORIES (31GB) のデータも追加で学習させている。
合計のpre-training用データセットのサイズは160GB以上で、BERTの10倍を超える。

### hyperparameterの見直し

BERTのbatch sizeは256であったが、batch sizeをより大きくすることで性能改善が見られた。RoBERTaでは8Kのbatch sizeを用いる (※実験では8Kよりも2Kの方がどの値も高かったのだが、そこに言及していない理由は不明)。

本文内でAdam epsilonの項に非常にsensitiveであったと述べている (※が、epsilonを変化させた実験は載っていない)。batch sizeを大きくした際にはβ\_2 = 0.98の値で安定してスコアが改善された。

### pre-trainingのステップをより多く

BERTがpre-trainingに100K step行ったのに対し、RoBERTaは160GBのデータを用いて100K, 300K, 500Kのstep数で実験を行い、stepが多くなるほど性能向上がみられることを確認した。このことからBERTはpre-trainingデータに対してunderfit状態であり、そのarchitectureは現在の性能よりも高い表現力を有していることを主張できる (と思う)。

### mask patternを動的に変更する

BERTのpre-trainingでは、40epochのうち、10回データにmaskを掛け直す作業を行っている。よってpre-training全体で、4回同じmask patternのデータをBERT modelは供給される計算になる。これは1つのインスタンスに対して10パターンのmaskを決め、それを静的にモデルに供給する方法である。

RoBERTaでは動的にmaskパターンを変えて学習を行っている。これはより多くのstep, より大きなデータセットでpre-trainingする際にはこの方法が重要となる、と主張している。

### next sentence predictionを廃止

next sentence prediction (NSP) の有用性を確認するため、4つの実験条件で比較する。

以下は作成したsentence-pairの有効なtoken数が条件によって異なる。それをbatch単位で揃えるため動的にbatch sizeを変更する処理をfull-sentencesとdoc-sentencesの条件で行っている。

- segment-pair

BERTの条件と同様。2つのsegment (segmentは複数のsentenceから成る可能性がある) を異なるsegment embeddingを用いて一つのベクトルにconcatしてNSP lossを使用する。

- sentence-pair+nsp

2つのsentenceをsegment embeddingを用いて一つのベクトルにconcatする。NSP lossを使用。

- full-sentences

一つ以上のdocumentから連続してサンプリングしたfull sentenceを用いて入力を作成する。入力内に別のdocumentが入った場合は、間にseparator tokenを挿入する。NSP lossを使用しない。

- doc-sentences

full-sentenceを、入力を一つのdocumentのみから構成されるように変更したもの。document終わり近くから作成した入力の場合、512 tokenに足らない入力が出来上がる場合がある。 (のでbatch sizeを動的に変更して対応する)

結果として、NSPが無いほうが性能が高かった。doc-sentencesが最も性能が高かったが、batch sizeを動的に扱うと計算コストがかかるので、full-sentencesの手法をRoBERTaには採用している。

### Text Encodingの方法を変更

BERTはBPEを使用してtokenを作成しているが、BPEの最小単位であるsubwordの単位はunicode文字であり、これはvocabularyのかなりの部分を占める。subwordの単位をbyteに変更することで、50K程度の語彙サイズでunknownラベルを使用せずとも入力文をencodeすることが可能になるようだ。

## 4. どうやって有効だと検証した？

GLUE, RACE, SQuADでSOTAを達成。GLUEデータセットではmulti-task learningを行っていない。SQuADでは追加のデータセットで学習を行っていない状態で性能が出ている。

## 5. 議論はある？

appendixのMRPCだけ違う動きをしているように見える (QQPも若干怪しいけど) 。
MRPCもQQPも、sentence pairが意味的に一致しているかを見るタスクだが。これnext sentence predictionを廃止したことによる影響だろうか？

大きなデータセット使ってBERTを訓練したら性能改善しました、という話がキモ？
